format: blkid content check, format argument, and verbose argument

eyz opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/linuxkit/linuxkit/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
1) Added new argument -force to the format utility. This will force formatting of the specified single device if it exists and is a block device
2) By default, the format package will no longer format the specified single device if content exists on that device
3) Added new blkid-based check for content, in addition to the existing check for partitions on autoformat candidate devices.
4) Cleanup of old unused code.
5) Refactoring of block device verification.
6) Added test/cases/040_packages/006_format_mount/005_by_device_force
7) Added document of new arguments to /docs/external-disk.md. Also sorted the arguments in the doc alphabetically and added them as bullets so they do not run together on the page.

**- How I did it**
- By default, the format package will no longer format the specified single device if content exists on that device
- Added new blkid-based check for content, in addition to the existing check for partitions on autoformat candidate devices.

**- How to verify it**
I have verified that the following will ignore auto-formatting devices other than `/dev/sda` -
```
  - name: format
    image: MYREGISTRY/format:<hash>
    command: ["/usr/bin/format", "-force", "/dev/sda"]
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**
![19575408_10211240830670443_6190493428953175106_o](https://user-images.githubusercontent.com/4591181/30764234-de8d1f52-9f9e-11e7-8c80-017b2be179bb.jpg)

justincormack commented:
Hmm, wouldn't this be more useful if you could specify multiple devices here? eg format sda or sdb but definitely dont mess with sdc?

eyz commented:
@justincormack, sure. I can do that

Something like `--devices sda,sdb,sdc`?

justincormack commented:
yeah that would be fine.

deitch commented:
> specify multiple devices here

I thought our philosophy here was to have single device, and just run the (lightweight) format container multiple times, once for each device, if wanted? 

I know we had discussed this before - cannot recall where at the moment - but that was the thrust. 

Personally, I have no issues with multiple, but I do recall that as the direction, at least then.

eyz commented:
@deitch, that was how I modeled my PR. Listing every device to format is nearly a full fdtab mapping

deitch commented:
> that was how I modeled my PR

Meaning one container of `format` for each device?

eyz commented:
@deitch one device per format

eyz commented:
Other thoughts? I can make another argument which specifies all of the devices which should be considered for auto formatting if not partitioned

deitch commented:
@justincormack whichever you think.

justincormack commented:
Hmm, well I was thinking of the multiple devices as a whitelist not a list to format, but I don't think that will work. I think this is fine as is.

eyz commented on pkg/format/format.go:
```diff
@@ -154,13 +155,17 @@ func findDrives() {
 	for _, d := range driveKeys {
-		drives[d] = true
+		if deviceVar == "" || deviceVar == d {
+			log.Printf("Considering auto format for device %s", d)
```
Logging to make the auto format candidate list visible to the end user

eyz replied:
I figured this would be helpful, as well.

eyz commented on pkg/format/format.go:
```diff
@@ -23,18 +23,40 @@ const (
 var (
 	labelVar  string
 	fsTypeVar string
+	deviceVar string
```
new var for --device argument

eyz commented on pkg/format/format.go:
```diff
@@ -23,18 +23,40 @@ const (
 func autoformat(label, fsType string) error {
 	var first string
 	for _, d := range driveKeys {
-		err := exec.Command("sfdisk", "-d", d).Run()
```
`sfdisk` check has been moved after new `blkid` check

eyz commented on pkg/format/format.go:
```diff
@@ -23,18 +23,40 @@ const (
 	var first string
 	for _, d := range driveKeys {
-		err := exec.Command("sfdisk", "-d", d).Run()
+		log.Printf("Considering auto format for device %s", d)
```
explicitly logging which devices we consider for auto-formatting to the console

eyz commented on pkg/format/format.go:
```diff
@@ -23,18 +23,40 @@ const (
+		out, err := exec.Command("blkid", d).Output()
+		if err == nil {
+			log.Printf("Skipping autoformat of %s. blkid returned: %s", d, out)
+			continue
```
if `blkid` returns with exitcode 0, then there was content detected on the device; skip the autoformat, and also include the uuid / label / type which was found, as applicable

eyz commented on pkg/format/format.go:
```diff
@@ -23,18 +23,40 @@ const (
+				if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+					// blkid exitcode 2 (from the non-busybox version) signifies the block device has no detectable content signatures
+					if status.ExitStatus() == 2 {
+						log.Printf("blkid did not find any existing content on %s.", d)
```
do not skip, as `blkid` exitcode 2 is returned when no content is detected (NOTE: only the `blkid` package from Alpine; the busybox `blkid` does not have this feature, so I install the alpine `blkid` package for this reason)

https://linux.die.net/man/8/blkid mentions this exitcode, should you like to see that man page

eyz commented on pkg/format/format.go:
```diff
@@ -23,18 +23,40 @@ const (
+						log.Printf("blkid did not find any existing content on %s.", d)
+					} else {
+						log.Printf("Skipping autoformat of %s. blkid %s returned: %s (exitcode %d)", d, out, status.ExitStatus())
+						continue
```
if we received an exitcode other than 0 or 2, then we assume there was a fatal error with the check, so skip formatting, and output the result from the `blkid` command; we should not proceed if `blkid` can't read the block device to be safe

eyz commented on pkg/format/format.go:
```diff
@@ -23,18 +23,40 @@ const (
+			}
+		}
+
+		err = exec.Command("sfdisk", "-d", d).Run()
```
the existing `sfdisk` check is performed if the `blkid` test encounters no issues, and empty content

eyz commented on pkg/format/format.go:
```diff
@@ -143,24 +165,31 @@ func findDrives() {
-	for _, d := range driveKeys {
-		drives[d] = true
+
+		driveKey := filepath.Join("/dev", d.Name())
```
factored the `driveKey` out of the prior `append` so I can compare it with the value of the `--device` argument (if applicable)

eyz commented on pkg/format/format.go:
```diff
@@ -143,24 +165,31 @@ func findDrives() {
+		// ignore if the deviceVar argument is defined and the driveKey doesn't match the deviceVar
+		if deviceVar != "" && driveKey != deviceVar {
+			continue
+		}
```
if the `--device` argument exists (non-empty) and the `driveKey` does not match the specified device from `--device`, then skip this device; this causes us to follow the above pattern of ignoring devices, but as applied to the new `--device` argument in this case

dave-tucker replied:
I assume that `--device` should be in the format `/dev/sda` then?

eyz commented on pkg/format/format.go:
```diff
@@ -143,24 +165,31 @@ func findDrives() {
+			continue
+		}
+
+		driveKeys = append(driveKeys, driveKey)
```
append the factored `driveKey` into `driveKeys` as it was before

eyz commented on pkg/format/format.go:
```diff
@@ -143,24 +165,31 @@ func findDrives() {
-	flag.StringVar(&labelVar, "label", "", "Disk label to apply")
 	flag.StringVar(&fsTypeVar, "type", "ext4", "Type of filesystem to create")
+	flag.StringVar(&labelVar, "label", "", "Disk label to apply")
+	flag.StringVar(&deviceVar, "device", "", "Device to format")
```
new `--device` argument, which is the explicit single device to allow to be formatted

dave-tucker replied:
nit: these were in alphabetical order before, it would be nice to keep them that way.
`device`, `label`, `type`.

eyz commented:
@dave-tucker you mentioned you'd like to take a look

dave-tucker commented on pkg/format/format.go:
```diff
@@ -23,18 +22,40 @@ const (
 			log.Printf("Partition table found on device %s. Skipping.", d)
 			continue
 		}
+
```
Why the additional whitespace?

eyz replied:
Personal style. I reverted since.

dave-tucker commented on pkg/format/format.go:
```diff
@@ -143,24 +164,31 @@ func findDrives() {
 		// this probably shouldn't be so hard
 		// but d.Mode()&os.ModeDevice == 0 doesn't work as expected
 		mode := d.Sys().(*syscall.Stat_t).Mode
+
```
and here...

dave-tucker commented on pkg/format/format.go:
```diff
@@ -143,24 +164,31 @@ func findDrives() {
 		if (mode & syscall.S_IFMT) != syscall.S_IFBLK {
 			continue
 		}
+
```
... and here ...

dave-tucker commented:
As another minor nit, could you please include a description of the changes in the git commit message? It's easier to see in the history. Also, the title might be better phrased as `format: Add device argument and blkid content check` - you can include more detail in the commit message.

eyz commented:
@dave-tucker is there preference about the device format? `/dev/sdX` or just `sdX`?

I was trying not to make large changes to the concepts of auto-discovery, but I'd be glad to have this flag skip findDrives.

So far: I had added spacing to split some of the checks into logical areas, to my eye when in the same function, but I removed the ones you mentioned. I also updated the commit log and the PR title.

eyz commented:
Dave Tucker and I agreed to change the arguments around a little. `--force` will be used to explicitly format the specified device, otherwise it will be checked for content and will be skipped if content is found.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "linuxkit-format-device-filter" git@github.com:eyz/linuxkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354126216
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

eyz commented on pkg/format/format.go:
```diff
@@ -21,36 +20,106 @@ const (
+func hasPartitions(d string) bool {
+	err := exec.Command("sfdisk", "-d", d).Run()
+	return err == nil
+}
```
factored out `sfdisk` partition check, as it is now used in multiple areas

eyz commented on pkg/format/Dockerfile:
```diff
@@ -11,6 +11,7 @@ RUN apk add --no-cache --initdb -p /out \
     musl \
     sfdisk \
     util-linux \
+    blkid \
```
the alpine package `blkid` is required
- the `busybox` version returns exitcode 0 when content is found
- the `blkid` package version returns exitcode 2 when content is found on the block device (distinct result)

eyz commented on pkg/format/format.go:
```diff
@@ -21,36 +20,106 @@ const (
-	if err := format(first, label, fsType); err != nil {
+	if err := format(first, label, fsType, false); err != nil {
 		return err
 	}
```
- false = do not force re-creation of partitions nor re-format
- true = clear partitions, re-partition, and format (if applicable)

eyz commented on pkg/format/format.go:
```diff
@@ -21,36 +20,106 @@ const (
+			}
+		}
+	}
+
```
support for `format` to optionally clear the partitions and re-partition via `forced` bool

eyz commented on pkg/format/format.go:
```diff
@@ -133,52 +203,91 @@ func format(d, label, fsType string) error {
+	mode := (*d).Sys().(*syscall.Stat_t).Mode
+	return (mode & syscall.S_IFMT) == syscall.S_IFBLK
+}
+
```
factored out the block device check, as it is used in multiple places now

eyz commented on pkg/format/format.go:
```diff
@@ -83,11 +152,12 @@ func format(d, label, fsType string) error {
+			if isBlockDevice(&stat) {
 				done = true
 				break
 			}
```
use new factored `isBlockDevice`

eyz commented on pkg/format/format.go:
```diff
@@ -133,52 +203,90 @@ func format(d, label, fsType string) error {
+				log.Printf("/dev/%s is not a block device", d.Name())
+			}
 			continue
 		}
```
- skip any devices which are not block devices
- using the new `isBlockDevice`

eyz commented on pkg/format/format.go:
```diff
@@ -133,52 +203,90 @@ func format(d, label, fsType string) error {
-	sort.Strings(driveKeys)
-	for _, d := range driveKeys {
-		drives[d] = true
-	}
```
not needed any longer

eyz commented on pkg/format/format.go:
```diff
@@ -133,52 +203,90 @@ func format(d, label, fsType string) error {
+	flag.BoolVar(&forceVar, "force", false, "Force format of specified single device (default false)")
 	flag.StringVar(&labelVar, "label", "", "Disk label to apply")
 	flag.StringVar(&fsTypeVar, "type", "ext4", "Type of filesystem to create")
+	flag.BoolVar(&verboseVar, "verbose", false, "Enable verbose output (default false)")
```
new `-force` and `-verbose` flags

eyz commented on pkg/format/format.go:
```diff
@@ -133,52 +203,90 @@ func format(d, label, fsType string) error {
+	}
+	// passed checks
+	return nil
 }
```
factored out device existence and block device checking

eyz commented on pkg/format/format.go:
```diff
@@ -133,52 +203,90 @@ func format(d, label, fsType string) error {
 
 	if flag.NArg() == 0 {
+		// auto-detect drives if a device to format is not explicitly specified
+		findDrives()
```
auto-detecting drives only if there are no explicit devices specified to consider singly

eyz commented on pkg/format/format.go:
```diff
@@ -133,52 +203,90 @@ func format(d, label, fsType string) error {
+			if err := autoformat(labelVar, fsTypeVar); err != nil {
+				log.Fatalf("%v", err)
+			}
+		}
```
if an explicit device was specified:
- ensure it is a block device, or error fatally
- if we are forcing (with `-force`) then format, or error fatally
- set `driveKeys` to an array with the specified device as the only element, which is still used by `autoformat`

rn commented on pkg/format/format.go:
```diff
@@ -21,36 +20,106 @@ const (
+			// blkid exitcode 2 (from the non-busybox version) signifies the block device has no detectable content signatures
+			if status.ExitStatus() == 2 {
+				if verboseVar {
+					log.Printf("blkid did not find any existing content on %s.", d)
```
why is this conditional on `verboseVar` but other output, like above, is not?

eyz replied:
@rn, which other line are you referring to?

eyz replied:
@rn This line is the verbose output, without a failure condition. The next one is a failure condition.

`blkid` exitcodes:
- `0` = content exists
- `2` = content does not exist
- other = error

eyz replied:
See below -

rn replied:
NM, just realised that you unconditionally print a message if, for whatever reason, the disk is not formatted and this case is not one of them, so all good

eyz replied:
The above isn't an error case. The one below is an error case. I only log verbose/debugging if `-verbose` is enabled, other than if it's an error (such as below)

eyz commented on pkg/format/format.go:
```diff
@@ -21,36 +20,106 @@ const (
+				// no content detected, but continue through all checks
+				isEmpty = true
+			} else {
+				return isEmpty, fmt.Errorf("Could not determine if %s was empty. blkid %s returned: %s (exitcode %d)", d, d, out, status.ExitStatus())
```
@rn was this the other line you meant? (see above)

eyz commented:
@rn I still owe @dave-tucker a new test case named `040_packages/006_format_mount/005_by_device_force` or such. I expect to work on that today, but I would also be curious to see if official CI passes. I have already confirmed that `rtf -x run` passes locally in qemu on a local build of this pkg.

eyz commented:
@dave-tucker @rn the docs have been updated, a new test has been written (and passes locally), and so I think it's up to @dave-tucker to give this a review, and then perhaps we can get this signed and pushed so it can be tested on LinuxKit CI.

eyz commented on pkg/format/format.go:
```diff
@@ -21,36 +20,133 @@ const (
+	time.Sleep(1 * time.Second)
+
+	return nil
+}
```
@rn just wanted to let you know that I factored this function out of `format` since you looked at this PR. It still passes tests locally.

rn commented:
@eyz could you squash the commits? I will then build/push the updated package and it would be great if you could then update the hashs in the YAML (in a separate commit) so that the CI has a chance to test with the update.

eyz commented:
@rn I just squashed that final commit.

rn commented:
Thanks! I've build and pushed the updated package for x86 and arm64. The new hash for the format package is `linuxkit/format:440a08cde1778aec60d9fb5de02d27f391cdfe37`. Could you create a *separate* commit which updates the YAMLs files. To do this, in the top-level directory run something like:
```
./scripts/update-component-sha.sh linuxkit/format:158d992b7bf7ab984100c697d7e72161ea7d7382 linuxkit/format:440a08cde1778aec60d9fb5de02d27f391cdfe37
```
and then commit the change.

eyz commented:
Updated `linuxkit/format' tag, and confirmed they match across the project -

```
$ grep -RF 'linuxkit/format' * | grep -Po 'linuxkit/format:\w+' | sort -u
linuxkit/format:440a08cde1778aec60d9fb5de02d27f391cdfe37

eyz commented:
@rn no problem. I dropped the hash update again just now. Let me know and I'll make the change ASAP

rn commented:
Images are pushed for arm64 and x86. Could you update the Dockerfile to use `linuxkit/alpine:d1778ee29f11475548f0e861f57005a84e82ba4e`

eyz commented:
linuxkit/alpine:d1778ee29f11475548f0e861f57005a84e82ba4e has been updated in pkg/format/Dockerfile, and re-squashed

rn commented on pkg/format/format.go:
```diff
@@ -21,36 +20,133 @@ const (
 	}
 
-	if err := format(first, label, fsType); err != nil {
+	if err := format(first, label, fsType, false); err != nil {
```
`golint` complains about this:
```
./format.go:97:2: redundant if ...; err != nil check, just return error instead.
sh: redundant: unknown operand
```

eyz replied:
Trying change to just
```
	return format(first, label, fsType, false)
```

rn commented on pkg/format/format.go:
```diff
@@ -21,36 +20,133 @@ const (
+			log.Printf("Considering auto format for device %s", d)
+		}
+		// break the loop with the first empty device we find
+		isEmpty, err := isEmptyDevice(d)
```
Since the code below needs changing, might as well refactor this as well to
```
if isEmpty, err := isEmptyDevice(d); err != nil {
  return err
} else {
  if isEmpty == true {
    first = d
    break
}
```
This limits the scope of `isEmpty`

eyz replied:
Doing so now. Good point.

rn commented on pkg/format/format.go:
```diff
@@ -21,36 +20,133 @@ const (
+	// wait for device
+	var done bool
+	for i := 0; i < timeout; i++ {
+		stat, err := os.Stat(awaitedDevice)
```
same as my comment here above.

eyz replied:
Trying change to just `
	return format(first, label, fsType, false)

eyz commented:
@rn can you run CI style checks on 91bcd4d3045c38df22e050ecaa6d5fc729e234ee ? I inverted and collapsed the err like you requested.

eyz commented:
@rn made your recommended changes. Please re-test golint

rn commented:
build and pushed x86/arm64 package. Please update YAML files with the new pacakge: `linuxkit/format:f180a74d878c8c0c86f6208f9311474c91452a79`

eyz commented:
Pushed tag update to linuxkit/format:f180a74d878c8c0c86f6208f9311474c91452a79

eyz commented:
@rn Sorry about that. I hadn't reverted my testing before I pushed. Confirmed no testing tags. Also, can you please redact the tag with my testing registry name in your comment? I am not supposed to refer to my organization per my agreement to contribute to open source.
